tests/Replication2/ReplicatedLog/CalcCommitIndexTest.cpp

```diff
@@ -31,39 +32,161 @@ using namespace arangodb::replication2::replicated_log;
+                                       CalculateCommitIndexOptions{2, 2, 3},
+                                       LogIndex{1}, LogIndex{50});
+  EXPECT_EQ(index, LogIndex{35});
+  EXPECT_TRUE(
+      std::holds_alternative<CommitFailReason::QuorumSizeNotReached>(reason.value));
+  EXPECT_EQ(quorum, (std::vector<ParticipantId>{"A", "C"}));
+}
+
+TEST_F(CalcCommitIndexTest, write_concern_3_3_participants) {
   auto participants = std::vector{
-      IndexParticipantPair{LogIndex{50}, "A"},
-      IndexParticipantPair{LogIndex{25}, "B"},
-      IndexParticipantPair{LogIndex{35}, "C"},
+      ParticipantStateTuple{LogIndex{50}, "A", {}},
+      ParticipantStateTuple{LogIndex{25}, "B", {}},
+      ParticipantStateTuple{LogIndex{35}, "C", {}},
   };
 
-  auto [index, reason] = algorithms::calculateCommitIndex(participants, 2, LogIndex{50});
-  EXPECT_EQ(index, LogIndex{35});
-  EXPECT_TRUE(std::holds_alternative<CommitFailReason::QuorumSizeNotReached>(reason.value));
+  auto [index, reason, quorum] =
+      algorithms::calculateCommitIndex(participants,
+                                       CalculateCommitIndexOptions{3, 3, 3},
+                                       LogIndex{1}, LogIndex{50});
+  EXPECT_EQ(index, LogIndex{25});
+  EXPECT_TRUE(
+      std::holds_alternative<CommitFailReason::QuorumSizeNotReached>(reason.value));
+
+  EXPECT_EQ(quorum, (std::vector<ParticipantId>{"A", "C", "B"}));
 }
 
-TEST_F(CalcCommitIndexTest, write_concern_3_3_participants) {
+TEST_F(CalcCommitIndexTest, includes_less_quorum_size) {
+  // Three participants but only two are included
+
+  auto participants = std::vector{ParticipantStateTuple{LogIndex{50}, "A", {ParticipantFlag::Excluded}},
+                                  ParticipantStateTuple{LogIndex{25}, "B", {}},
+                                  ParticipantStateTuple{LogIndex{35}, "C", {}}};
+
+  auto [index, reason, quorum] =
+      algorithms::calculateCommitIndex(participants,
+                                       CalculateCommitIndexOptions{3, 3, 3},
+                                       LogIndex{1}, LogIndex{50});
+  EXPECT_EQ(index, LogIndex{1});
+  EXPECT_TRUE(
+      std::holds_alternative<CommitFailReason::QuorumSizeNotReached>(reason.value));
+  EXPECT_EQ(quorum, (std::vector<ParticipantId>{}));
+}
+
+TEST_F(CalcCommitIndexTest, excluded_and_forced) {
+  // One participant is excluded *and* forced, which means we cannot make any progress
+
+  auto participants = std::vector{
+      ParticipantStateTuple{LogIndex{50}, "A", {}},
+      ParticipantStateTuple{LogIndex{25}, "B", {ParticipantFlag::Excluded, ParticipantFlag::Forced}},
+      ParticipantStateTuple{LogIndex{35}, "C", {}},
+  };
+
+  auto [index, reason, quorum] =
+      algorithms::calculateCommitIndex(participants,
+                                       CalculateCommitIndexOptions{2, 2, 3},
+                                       LogIndex{1}, LogIndex{50});
+  EXPECT_EQ(index, LogIndex{1});
```

Comment:
I would expect this to be 25. `B` has received entry 25, so the force attribute does not prevent it. However, it is excluded, and thus not allowed to be part of the quorum for entry 25. But A and C could form a quorum for index 25. Write concern is 2.